src/libraries/System.Data.Common/src/System/Xml/BaseTreeIterator.cs

```diff
@@ -47,10 +52,16 @@ internal bool NextRightRowElement()
         }
 
         // Returns true if the current node is on a row element (head of a region)
+        [MemberNotNullWhen(true, nameof(CurrentNode))]
         internal bool OnRowElement()
         {
-            XmlBoundElement be = CurrentNode as XmlBoundElement;
-            return (be != null) && (be.Row != null);
+            XmlBoundElement? be = CurrentNode as XmlBoundElement;
+            if ((be != null) && (be.Row != null))
+            {
+                Debug.Assert(CurrentNode != null);
+                return true;
+            }
+            return false;
```

Comment:
I would expect a simpler change would still satisfy the compiler:
```C#
return CurrentNode is XmlBoundElement be && be.Row != null;
```
Does it not?

Reply:
It does. For some reason, this doesn't work:

```C#
XmlBoundElement? be = CurrentNode as XmlBoundElement;
return ((be != null) && (be.Row != null));
```

Fixed

Reply:
> For some reason, this doesn't work

The compiler doesn't track the relationship between variables for nullability.